cache/src/main/kotlin/com/dropbox/android/external/cache4/RealCache.kt

```diff
@@ -0,0 +1,328 @@
+ *
+ * Size-based evictions are enabled by specifying [maxSize]. When the size of the cache entries grows
+ * beyond [maxSize], least recently accessed entries will be evicted.
+ */
+internal class RealCache<Key : Any, Value : Any>(
+    val expireAfterWriteNanos: Long,
+    val expireAfterAccessNanos: Long,
+    val maxSize: Long,
+    val concurrencyLevel: Int,
+    val clock: Clock
+) : Cache<Key, Value> {
+
+    /**
+     * A map holding the current cache entries.
+     */
+    private val cacheEntries: MutableMap<Key, CacheEntry<Key, Value>> = ConcurrentHashMap(
+        INITIAL_CAPACITY, LOAD_FACTOR, concurrencyLevel
+    )
+
+    /**
+     * A queue of unique cache entries ordered by write time.
+     * Used for performing write-time based cache expiration.
+     */
+    private val writeQueue: MutableSet<CacheEntry<Key, Value>>?
+
+    /**
+     * A queue of unique cache entries ordered by access time.
+     * Used for performing both write-time and read-time based cache expiration
+     * as well as size-based eviction.
+     *
+     * Note that a write is also considered an access.
+     */
+    private val accessQueue: MutableSet<CacheEntry<Key, Value>>?
+
+    /**
+     * Whether to perform size based evictions.
+     */
+    private val evictsBySize = maxSize >= 0
+
+    /**
+     * Whether to perform write-time based expiration.
+     */
+    private val expiresAfterWrite = expireAfterWriteNanos > 0
+
+    /**
+     * Whether to perform access-time (both read and write) based expiration.
+     */
+    private val expiresAfterAccess = expireAfterAccessNanos > 0
+
+    /**
+     * A key-based synchronizer for running cache loaders.
+     */
+    private val loadersSynchronizer = KeyedSynchronizer<Key>()
+
+    init {
+        // writeQueue is required if write expiry is enabled
+        writeQueue = takeIf { expiresAfterWrite }?.let {
+            Collections.synchronizedSet(ReorderingLinkedHashSet())
+        }
+
+        // accessQueue is required if either read expiry is enabled or size-based eviction is enabled
+        accessQueue = takeIf { expiresAfterAccess || evictsBySize }?.let {
+            Collections.synchronizedSet(ReorderingLinkedHashSet())
+        }
+    }
+
+    override fun get(key: Key): Value? {
+        val nowNanos = clock.currentTimeNanos
+        return cacheEntries[key]?.let {
+            if (it.isExpired(nowNanos)) {
+                // clean up expired entries and return null
+                expireEntries(nowNanos)
+                null
+            } else {
+                // update eviction metadata
+                recordRead(it, nowNanos)
+                // if entry's access time has been updated, perform an update to the map to create
+                // a memory barrier for synchronizing subsequent access to the map from other threads.
+                if (expiresAfterAccess) {
+                    cacheEntries[key] = it
```

Comment:
in retrospect, it's probably better to not re-add entries to the map but make the `var`s volatile. the (rare) race condition we have not is:
thread 1 reads value1 and calls recordRead
thread 2 evicts value1 from cache
thread 2 adds value2 to the map
thread 1 resumes and re-adds value1 to the map, replacing the newer value2.

let's avoid this by removing all the `cacheEntry[key] = it` calls in the map and marks the `var`s in cache entry as `@Volatile`